### ISSUE TITLE
Rebase vs Merge

### ISSUE THREAD

**Issue by kalimar** [Terrastories/terrastories#290]
Today I spent some time with one of our contributors rebasing their branch before merging with masters. That's the philosophy I've always practiced at work, on a team of only a few people. After going through this exercise it occurred to me that when many people are contributing to the project, that might be less realistic and perhaps we just need to rely on automated tests to confirm if a branch is ready to be merged. Otherwise, everyone will essentially be forced into a queue and "wait their turn" to rebase and push their branch and then merge with master.

Thoughts? I don't contribute to enough other open-source projects to know what's the most common approach. cc @csexton can you weigh in?

**Comment by namel3ss** [comment]
I am used to the Rebase approach. At my job there are around 10-15 people working concurrently on the same project. We still use rebase. It might be a bit annoying to constantly have to rebase you branch because something was merged in master. But it doesn't make a big difference what approach you are using. If there are no conflicts the pull request will be green anyway and someone can merge it. And if not green, no matter if you use merge or rebase, someone still has to solve the conflicts them before merging the PR.

**Comment by kalimar** [comment]
I've been in the same boat - we all used rebase at my last job. But I'm leaning towards making it much easier for everyone and go with merge. Which makes having tests all the more important imo. I appreciate you weighing in. Having a linear history isn't crucial for us and worth the extra work.

**Comment by mirandawang** [comment]
The main concern I have with not enforcing a rebase is that we don't currently have a super robust test suite. This might mean that merges could often break master. But I also see what you mean about forcing everyone to be in a "queue" which can be annoying. Currently, only a few people have the permissions to merge back into the master terrastories branch. Maybe that could be a responsibility of whoever's reviewing to merge to fix conflicts before they do and make sure master is still working? Is that too much? A reviewer could approve a PR after they've tested it, fix the conflicts, test it (manually or however) and then merge it in if it's working. If it's not working, they could comment what bugs were introduced after the rebase. Not sure how to share the merge conflict resolutions though, since that would require a force push.

**Comment by namel3ss** [comment]
I would rather much delegate conflicts to the developer of the PR. Otherwise the reviewer may end up into situation to resolve conflicts in code that may not be 100% familiar for him, and thus may again introduce bugs.

Actually this has been a topic in my current team from sometime now. And there is no easy solution. Because every solution has its own pros and cons. For example, I often check my PRs for conflicts maybe daily or so. But it is easier when you work full-time on a project. Here, the case might be different because of the time difference and the different availability of people. Also for the reviewer to fix conflicts, he needs write access to the fork. If it is a local branch that is not an issue.

In conclusion, for me contributing does not end after you submit a PR. I am responsible for the PR until it gets merged and conflicts are my responsibility to fix. Owners and maintainers should not be burdened with that.

**Comment by mirandawang** [comment]
@namel3ss Sorry for the super delayed response! :sweat_smile: 

I think what you're saying makes a lot of sense about the developer of the PR knowing how to resolve the conflicts best.

In the case of a merge, does that mean that the developer would be resolving conflicts such that it is not going to break the master branch? If so then I think that would mean the same safety to me as a rebase. However, isn't it still the case that if more changes to master come, they would have to do another merge (as they would have to, with a rebase as well) to get up to date again? And there would still be a queue? Excuse my not-too-sophisticated knowledge of git merge/rebases :)
